src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,1664 @@
+      /* Offset of the map entry into the log entry's buffer */
+      uint64_t map_entry_buffer_offset = entry_image_extent.first - entry.log_entry->ram_entry.image_offset_bytes;
+      /* Offset into the log entry buffer of this read hit */
+      uint64_t read_buffer_offset = map_entry_buffer_offset + entry_offset;
+      /* Create buffer object referring to pmem pool for this read hit */
+      WriteLogEntry *log_entry = entry.log_entry;
+      ldout(cct, 5) << "adding reader: log_entry=" << *log_entry << dendl;
+      log_entry->add_reader();
+      buffer::raw *hit_buf =
+	buffer::claim_buffer(entry_hit_length,
+			     (char*)(log_entry->pmem_buffer + read_buffer_offset),
+			     make_deleter([this, log_entry]
+					  {
+					    CephContext *cct = m_image_ctx.cct;
+					    ldout(cct, 5) << "removing reader: log_entry="
+							  << *log_entry << dendl;
+					    log_entry->remove_reader();
+					  }));
+      /* Add hit extent to read extents */
+      ImageExtentBuf hit_extent_buf(hit_extent, hit_buf);
+      read_ctx->m_read_extents.push_back(hit_extent_buf);
+      /* Exclude RWL hit range from buffer and extent */
+      extent_offset += entry_hit_length;
+      ldout(cct, 20) << entry << dendl;
+    }
+    /* If the last map entry didn't consume the entire image extent ... */
+    if (extent.second > extent_offset) {
+      /* ... add the rest of this extent to miss extents */
+      uint64_t miss_extent_start = extent.first + extent_offset;
+      uint64_t miss_extent_length = extent.second - extent_offset;
+      ImageCache::Extent miss_extent(miss_extent_start, miss_extent_length);
+      read_ctx->m_miss_extents.push_back(miss_extent);
+      /* Add miss range to read extents */
+      ImageExtentBuf miss_extent_buf(miss_extent);
+      read_ctx->m_read_extents.push_back(miss_extent_buf);
+      extent_offset += miss_extent_length;
+    }
+  }
+  
+  ldout(cct, 5) << "miss_extents=" << read_ctx->m_miss_extents << ", "
+		<< "miss_bl=" << read_ctx->m_miss_bl << dendl;
+
+  if (read_ctx->m_miss_extents.empty()) {
+    /* All of this read comes from RWL */
+    read_ctx->complete(0);
+  } else {
+    /* Pass the read misses on to the layer below RWL */
+    if (m_image_ctx.persistent_cache_enabled) {
+      m_image_cache.aio_read(std::move(read_ctx->m_miss_extents), &read_ctx->m_miss_bl, fadvise_flags, read_ctx);
+    } else {
+      m_image_writeback.aio_read(std::move(read_ctx->m_miss_extents), &read_ctx->m_miss_bl, fadvise_flags, read_ctx);
+    }
+  }
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::detain_guarded_request(GuardedRequest &&req)
+{
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << dendl;
+  
+  BlockGuardCell *cell;
+  int r = m_write_log_guard.detain({req.first_block_num, req.last_block_num},
+				   &req, &cell);
+  if (r < 0) {
+    lderr(cct) << "failed to detain guarded request: " << cpp_strerror(r)
+		   << dendl;
+    m_image_ctx.op_work_queue->queue(req.on_guard_acquire, r);
```

Comment:
Yeah, I think it will. Worse, I think it will leak and not complete the actual user request.

I really don't think this detain operation can fail, assuming I'm looking at the right implementation of detain(). This if() should probably just be assert(r >= 0).

Reply:
This was turned into an assert